controllers/runnerjob_controller.go

```diff
@@ -0,0 +1,148 @@
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/actions-runner-controller/actions-runner-controller/api/v1alpha1"
+	"github.com/go-logr/logr"
+	batchv1 "k8s.io/api/batch/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// RunnerJobReconciler reconciles a RunnerJob object
+type RunnerJobReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+const jobOwnerKey string = ".metadata.controller"
+
+var apiGVStr = v1alpha1.GroupVersion.String()
+
+//+kubebuilder:rbac:groups=actions.summerwind.dev,resources=runnerjobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=actions.summerwind.dev,resources=runnerjobs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=batch,resources=jobs/status,verbs=get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the RunnerJob object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *RunnerJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("runnerjob", req.NamespacedName)
+
+	var runnerJob v1alpha1.RunnerJob
+	if err := r.Get(ctx, req.NamespacedName, &runnerJob); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !runnerJob.ObjectMeta.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	var childJobs batchv1.JobList
+	if err := r.List(ctx, &childJobs, client.InNamespace(req.Namespace), client.MatchingFields{jobOwnerKey: req.Name}); err != nil {
+		log.Error(err, "unable to list child Jobs")
+		return ctrl.Result{}, err
+	}
+
+	switch len(childJobs.Items) {
+	case 0:
+		if err := r.createJob(ctx, runnerJob); err != nil {
+			log.Info("failed to create a job from a RunnerJob, requeueing in 30s", runnerJob.Name, err.Error())
+			return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+		}
+	case 1:
+		childJob := childJobs.Items[0]
+		if childJob.Status.Active > 0 {
+			return ctrl.Result{}, nil
+		}
+
+		if childJob.Status.Succeeded > 0 {
+			if err := r.Delete(ctx, &runnerJob); err != nil {
+				log.Info("failed to delete runner job", "RunnerJob", runnerJob.Name)
+				return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+			}
+			return ctrl.Result{}, nil
+		}
+
+		log.Info("more than one child pods failed", "RunnerJob", runnerJob.Name, "Job", childJob.Name)
+	default:
```

Comment:
TEST: Does returning {}, err in Reconcile trigger reconciliation again?

Reply:
Here is how the controller runtime will handle the difference between returning an error and re-queuing.

https://github.com/kubernetes-sigs/controller-runtime/blob/f0351217e9e026533aece74d054b23cae5441659/pkg/internal/controller/controller.go#L318-L343

If we only want to tell the user that we can't reconcile, I think we need to log the error in our controller and `return {}, nil` to prevent reconciliation from retrying.